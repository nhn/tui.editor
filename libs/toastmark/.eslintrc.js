module.exports = {
  root: true,
  env: {
    browser: true,
    commonjs: true,
    node: true,
    jest: true
  },
  parser: '@typescript-eslint/parser',
  plugins: ['prettier', '@typescript-eslint'],
  extends: [
    'tui/es6',
    'plugin:@typescript-eslint/recommended',
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended'
  ],
  parserOptions: {
    parser: 'typescript-eslint-parser'
  },
  rules: {
    'prefer-destructuring': 0,
    'newline-before-return': 0,
    'padding-line-between-statements': 0,
    'lines-between-class-members': 0,
    'no-useless-constructor': 0,
    'default-param-last': 1,
    'no-dupe-class-members': 0,
    'no-cond-assign': 0,
    'no-plusplus': 0,
    eqeqeq: 0,
    'no-empty': 0,
    'no-shadow': 0,
    'max-depth': 0,
    'max-nested-callbacks': 0,
    'no-useless-escape': 0,
    'no-control-regex': 0,
    'no-constant-condition': 0,
    'accessor-pairs': 0,
    'require-jsdoc': 0,
    'no-undefined': 0,
    'no-new': 0,
    'spaced-comment': 0,
    'no-lonely-if': 0,
    'no-nested-ternary': 0,
    curly: 2,
    complexity: 0,
    '@typescript-eslint/no-non-null-assertion': 0,
    '@typescript-eslint/explicit-function-return-type': 0,
    '@typescript-eslint/no-explicit-any': 0,
    '@typescript-eslint/no-triple-slash-reference': 0,
    '@typescript-eslint/no-object-literal-type-assertion': 0,
    '@typescript-eslint/no-use-before-define': 0,
    '@typescript-eslint/triple-slash-reference': 0,
    '@typescript-eslint/interface-name-prefix': 0,
    '@typescript-eslint/no-useless-constructor': 2,
    '@typescript-eslint/ban-ts-ignore': 0,
    '@typescript-eslint/no-empty-function': 0
  }
};
