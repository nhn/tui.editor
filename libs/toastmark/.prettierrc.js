module.exports = {
  "printWidth" : 100,
  "singleQuote" : true
}
