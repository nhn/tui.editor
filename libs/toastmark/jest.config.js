module.exports = {
  preset: 'ts-jest',
  testMatch: ['**/__test__/*.spec.ts'],
  testEnvironment: 'node'
};
