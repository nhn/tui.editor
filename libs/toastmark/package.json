{
  "name": "@toast-ui/toastmark",
  "version": "0.0.1-alpha.1",
  "description": "ToastMark - Incremental markdown parser extended from CommonMark.js",
  "scripts": {
    "lint": "eslint .",
    "test:types": "tsc",
    "test": "jest --watch",
    "test:ci": "jest",
    "serve": "snowpack dev",
    "serve:ie": "webpack serve",
    "build": "rollup -c && webpack build"
  },
  "types": "types/index.d.ts",
  "files": [
    "dist",
    "types"
  ],
  "keywords": [
    "markdown",
    "parser"
  ],
  "author": "NHN",
  "main": "dist/toastmark.js",
  "module": "dist/esm/index.js",
  "license": "MIT",
  "devDependencies": {
    "@types/codemirror": "5.60.5",
    "@types/mdurl": "^1.0.2",
    "codemirror": "^5.51.0",
    "entities": "^4.3.0",
    "mdurl": "^1.0.1"
  }
}
