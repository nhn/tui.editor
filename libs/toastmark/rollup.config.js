import typescript from '@rollup/plugin-typescript';
import commonjs from '@rollup/plugin-commonjs';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import json from '@rollup/plugin-json';

export default [
  {
    input: 'src/index.ts',
    output: {
      dir: 'dist/esm',
      format: 'es',
      sourcemap: false,
    },
    plugins: [typescript(), commonjs(), nodeResolve(), json()],
  },
];
