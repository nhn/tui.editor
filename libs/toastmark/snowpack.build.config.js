// eslint-disable-next-line @typescript-eslint/no-var-requires
/** @type {import("snowpack").SnowpackUserConfig } */
module.exports = {
  mount: {
    src: '/',
  },
  exclude: ['**/__sample__/**/*', '**/__test__/**/*'],
  buildOptions: {
    out: 'dist/esm',
  },
  optimize: {
    entrypoints: ['./src/index.ts'],
    bundle: true,
  },
  alias: {
    '@': './src',
  },
};
