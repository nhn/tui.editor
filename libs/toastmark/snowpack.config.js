/** @type {import("snowpack").SnowpackUserConfig } */
module.exports = {
  mount: {
    demo: '/',
    src: '/dist',
  },
  devOptions: {
    port: 8000,
  },
  alias: {
    '@t': './types',
  },
};
