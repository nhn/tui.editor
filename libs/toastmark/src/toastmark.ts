import { Parser, Options } from './commonmark/blocks';
import {
  BlockNode,
  isList,
  removeAllNode,
  removeNodeById,
  Node,
  SourcePos,
  isRefDef,
  RefDefNode,
  isTable
} from './commonmark/node';
import {
  removeNextUntil,
  getChildNodes,
  insertNodesBefore,
  prependChildNodes,
  updateNextLineNumbers,
  findChildNodeAtLine,
  findFirstNodeAtLine,
  findNodeAtPosition,
  findNodeById,
  invokeNextUntil,
  isUnlinked
} from './nodeHelper';
import { reBulletListMarker, reOrderedListMarker } from './commonmark/blockStarts';
import { iterateObject, omit, isEmptyObj } from './helper';
import { isBlank } from './commonmark/blockHelper';

const reLineEnding = /\r\n|\n|\r/;

export type Position = [number, number];

export type Range = [Position, Position];

type EventName = 'change';

type EventHandlerMap = {
  [key in EventName]: Function[];
};

type RemovedNodeRange = {
  id: [number, number];
  line: [number, number];
};

interface EditResult {
  nodes: BlockNode[];
  removedNodeRange: RemovedNodeRange | null;
}

type ParseResult = EditResult & { nextNode: Node | null };
type RefDefState = {
  id: number;
  destination: string;
  title: string;
  unlinked: boolean;
  sourcepos: SourcePos;
};

export type RefMap = {
  [k: string]: RefDefState;
};

export type RefLinkCandidateMap = {
  [k: number]: {
    node: BlockNode;
    refLabel: string;
  };
};

export type RefDefCandidateMap = {
  [k: number]: RefDefNode;
};

function canBeContinuedListItem(lineText: string) {
  const spaceMatch = lineText.match(/^[ \t]+/);
  if (spaceMatch && (spaceMatch[0].length >= 2 || /\t/.test(spaceMatch[0]))) {
    return true;
  }

  const leftTrimmed = spaceMatch ? lineText.slice(spaceMatch.length) : lineText;
  return reBulletListMarker.test(leftTrimmed) || reOrderedListMarker.test(leftTrimmed);
}

function canBeContinuedTableBody(lineText: string) {
  return !isBlank(lineText) && lineText.indexOf('|') !== -1;
}

export function createRefDefState(node: RefDefNode) {
  const { id, title, sourcepos, dest } = node;
  return {
    id,
    title,
    sourcepos: sourcepos!,
    unlinked: false,
    destination: dest
  };
}

export class ToastMark {
  public lineTexts: string[];
  private parser: Parser;
  private root: BlockNode;
  private eventHandlerMap: EventHandlerMap;
  private refMap: RefMap;
  private refLinkCandidateMap: RefLinkCandidateMap;
  private refDefCandidateMap: RefDefCandidateMap;
  private useReferenceDefinition: boolean;

  constructor(contents?: string, options?: Partial<Options>) {
    this.refMap = {};
    this.refLinkCandidateMap = {};
    this.refDefCandidateMap = {};
    this.useReferenceDefinition = !!options?.useReferenceDefinition;
    this.parser = new Parser(options);
    this.parser.setRefMaps(this.refMap, this.refLinkCandidateMap, this.refDefCandidateMap);
    this.eventHandlerMap = { change: [] };

    contents = contents || '';
    this.lineTexts = contents.split(reLineEnding);
    this.root = this.parser.parse(contents);
  }

  private updateLineTexts(startPos: Position, endPos: Position, newText: string) {
    const [startLine, startCol] = startPos;
    const [endLine, endCol] = endPos;
    const newLines = newText.split(reLineEnding);
    const newLineLen = newLines.length;
    const startLineText = this.lineTexts[startLine - 1];
    const endLineText = this.lineTexts[endLine - 1];
    newLines[0] = startLineText.slice(0, startCol - 1) + newLines[0];
    newLines[newLineLen - 1] = newLines[newLineLen - 1] + endLineText.slice(endCol - 1);

    const removedLineLen = endLine - startLine + 1;
    this.lineTexts.splice(startLine - 1, removedLineLen, ...newLines);

    return newLineLen - removedLineLen;
  }

  private updateRootNodeState() {
    if (this.lineTexts.length === 1 && this.lineTexts[0] === '') {
      this.root.lastLineBlank = true;
      this.root.sourcepos = [[1, 1] as Position, [1, 0] as Position];
      return;
    }

    if (this.root.lastChild) {
      this.root.lastLineBlank = (this.root.lastChild as BlockNode).lastLineBlank;
    }

    const { lineTexts } = this;
    let idx = lineTexts.length - 1;
    while (lineTexts[idx] === '') {
      idx -= 1;
    }
    if (lineTexts.length - 2 > idx) {
      idx += 1;
    }

    this.root.sourcepos![1] = [idx + 1, lineTexts[idx].length];
  }

  private replaceRangeNodes(
    startNode: BlockNode | null,
    endNode: BlockNode | null,
    newNodes: BlockNode[]
  ) {
    if (!startNode) {
      if (endNode) {
        insertNodesBefore(endNode, newNodes);
        removeNodeById(endNode.id);
        endNode.unlink();
      } else {
        prependChildNodes(this.root, newNodes);
      }
    } else {
      insertNodesBefore(startNode, newNodes);
      removeNextUntil(startNode, endNode!);
      [startNode.id, endNode!.id].forEach(id => removeNodeById(id));
      startNode.unlink();
    }
  }

  private getNodeRange(startPos: Position, endPos: Position) {
    const startNode = findChildNodeAtLine(this.root, startPos[0]);
    let endNode = findChildNodeAtLine(this.root, endPos[0]);

    // extend node range to include a following block which doesn't have preceding blank line
    if (endNode && endNode.next && endPos[0] + 1 === endNode.next.sourcepos![0][0]) {
      endNode = endNode.next;
    }

    return [startNode, endNode] as [BlockNode, BlockNode];
  }

  private trigger(eventName: EventName, param: any) {
    this.eventHandlerMap[eventName].forEach(handler => {
      handler(param);
    });
  }

  private extendEndLine(line: number) {
    while (this.lineTexts[line] === '') {
      line += 1;
    }
    return line;
  }

  private parseRange(
    startNode: BlockNode | null,
    endNode: BlockNode | null,
    startLine: number,
    endLine: number
  ) {
    // extends starting range if the first node can be a continued list item
    if (
      startNode &&
      startNode.prev &&
      ((isList(startNode.prev) && canBeContinuedListItem(this.lineTexts[startLine - 1])) ||
        (isTable(startNode.prev) && canBeContinuedTableBody(this.lineTexts[startLine - 1])))
    ) {
      startNode = startNode.prev;
      startLine = startNode.sourcepos![0][0];
    }

    const editedLines = this.lineTexts.slice(startLine - 1, endLine);
    const root = this.parser.partialParseStart(startLine, editedLines);

    // extends ending range if the following node can be a continued list item
    let nextNode = endNode ? endNode.next : this.root.firstChild;
    while (
      root.lastChild &&
      isList(root.lastChild) &&
      nextNode &&
      (nextNode.type === 'list' || nextNode.sourcepos![0][1] >= 2)
    ) {
      const newEndLine = this.extendEndLine(nextNode.sourcepos![1][0]);
      this.parser.partialParseExtends(this.lineTexts.slice(endLine, newEndLine));

      if (!startNode) {
        startNode = endNode;
      }
      endNode = nextNode as BlockNode;
      endLine = newEndLine;
      nextNode = nextNode.next;
    }

    this.parser.partialParseFinish();

    const newNodes = getChildNodes(root)! as BlockNode[];
    return { newNodes, extStartNode: startNode, extEndNode: endNode };
  }

  private getRemovedNodeRange(
    extStartNode: BlockNode | null,
    extEndNode: BlockNode | null
  ): RemovedNodeRange | null {
    if (
      !extStartNode ||
      (extStartNode && isRefDef(extStartNode)) ||
      (extEndNode && isRefDef(extEndNode))
    ) {
      return null;
    }
    return {
      id: [extStartNode.id, extEndNode!.id],
      line: [extStartNode.sourcepos![0][0] - 1, extEndNode!.sourcepos![1][0] - 1]
    };
  }

  private markDeletedRefMap(extStartNode: BlockNode | null, extEndNode: BlockNode | null) {
    if (!isEmptyObj(this.refMap)) {
      const markDeleted = (node: BlockNode) => {
        if (isRefDef(node)) {
          const refDefState = this.refMap[node.label];
          if (refDefState && node.id === refDefState.id) {
            refDefState.unlinked = true;
          }
        }
      };
      if (extStartNode) {
        invokeNextUntil(markDeleted, extStartNode.parent!, extEndNode);
      }
      if (extEndNode) {
        invokeNextUntil(markDeleted, extEndNode);
      }
    }
  }

  private replaceWithNewRefDefState(nodes: BlockNode[]) {
    if (!isEmptyObj(this.refMap)) {
      const replaceWith = (node: BlockNode) => {
        if (isRefDef(node)) {
          const { label } = node;
          const refDefState = this.refMap[label];
          if (!refDefState || refDefState.unlinked) {
            this.refMap[label] = createRefDefState(node);
          }
        }
      };
      nodes.forEach(node => {
        invokeNextUntil(replaceWith, node);
      });
    }
  }

  private replaceWithRefDefCandidate() {
    if (!isEmptyObj(this.refDefCandidateMap)) {
      iterateObject(this.refDefCandidateMap, (_, candidate) => {
        const { label, sourcepos } = candidate;
        const refDefState = this.refMap[label];

        if (
          !refDefState ||
          refDefState.unlinked ||
          refDefState.sourcepos[0][0] > sourcepos![0][0]
        ) {
          this.refMap[label] = createRefDefState(candidate);
        }
      });
    }
  }

  private getRangeWithRefDef(
    startLine: number,
    endLine: number,
    startNode: BlockNode,
    endNode: BlockNode,
    lineDiff: number
  ) {
    if (this.useReferenceDefinition && !isEmptyObj(this.refMap)) {
      const prevNode = findChildNodeAtLine(this.root, startLine - 1);
      const nextNode = findChildNodeAtLine(this.root, endLine + 1);

      if (prevNode && isRefDef(prevNode) && prevNode !== startNode && prevNode !== endNode) {
        startNode = prevNode;
        startLine = startNode.sourcepos![0][0];
      }

      if (nextNode && isRefDef(nextNode) && nextNode !== startNode && nextNode !== endNode) {
        endNode = nextNode;
        endLine = this.extendEndLine(endNode.sourcepos![1][0] + lineDiff);
      }
    }

    return [startNode, endNode, startLine, endLine] as const;
  }

  private parse(startPos: Position, endPos: Position, lineDiff = 0): ParseResult {
    const range = this.getNodeRange(startPos, endPos);
    const [startNode, endNode] = range;
    const startLine = startNode ? Math.min(startNode.sourcepos![0][0], startPos[0]) : startPos[0];
    const endLine = this.extendEndLine(
      (endNode ? Math.max(endNode.sourcepos![1][0], endPos[0]) : endPos[0]) + lineDiff
    );

    const parseResult = this.parseRange(
      ...this.getRangeWithRefDef(startLine, endLine, startNode, endNode, lineDiff)
    );
    const { newNodes, extStartNode, extEndNode } = parseResult;
    const removedNodeRange = this.getRemovedNodeRange(extStartNode, extEndNode);

    const nextNode = extEndNode ? extEndNode.next : this.root.firstChild;

    if (this.useReferenceDefinition) {
      this.markDeletedRefMap(extStartNode, extEndNode);
      this.replaceRangeNodes(extStartNode, extEndNode, newNodes);
      this.replaceWithNewRefDefState(newNodes);
    } else {
      this.replaceRangeNodes(extStartNode, extEndNode, newNodes);
    }

    return { nodes: newNodes, removedNodeRange, nextNode };
  }

  private parseRefLink() {
    const result: EditResult[] = [];

    if (!isEmptyObj(this.refMap)) {
      iterateObject(this.refMap, (label, value) => {
        if (value.unlinked) {
          delete this.refMap[label];
        }
        iterateObject(this.refLinkCandidateMap, (_, candidate) => {
          const { node, refLabel } = candidate;
          if (refLabel === label) {
            result.push(this.parse(node.sourcepos![0], node.sourcepos![1]));
          }
        });
      });
    }

    return result;
  }

  private removeUnlinkedCandidate() {
    if (!isEmptyObj(this.refDefCandidateMap)) {
      [this.refLinkCandidateMap, this.refDefCandidateMap].forEach(candidateMap => {
        iterateObject(candidateMap, id => {
          if (isUnlinked(id)) {
            delete candidateMap[id];
          }
        });
      });
    }
  }

  public editMarkdown(startPos: Position, endPos: Position, newText: string) {
    const lineDiff = this.updateLineTexts(startPos, endPos, newText);
    const parseResult = this.parse(startPos, endPos, lineDiff);
    const editResult: EditResult = omit(parseResult, 'nextNode');

    updateNextLineNumbers(parseResult.nextNode, lineDiff);
    this.updateRootNodeState();

    let result: EditResult[] = [editResult];

    if (this.useReferenceDefinition) {
      this.removeUnlinkedCandidate();
      this.replaceWithRefDefCandidate();
      result = result.concat(this.parseRefLink());
    }

    this.trigger('change', result);

    return result;
  }

  public getLineTexts() {
    return this.lineTexts;
  }

  public getRootNode() {
    return this.root;
  }

  public findNodeAtPosition(pos: Position) {
    const node = findNodeAtPosition(this.root, pos);
    if (!node || node === this.root) {
      return null;
    }
    return node;
  }

  public findFirstNodeAtLine(line: number) {
    return findFirstNodeAtLine(this.root, line);
  }

  public on(eventName: EventName, callback: Function) {
    this.eventHandlerMap[eventName].push(callback);
  }

  public off(eventName: EventName, callback: Function) {
    const handlers = this.eventHandlerMap[eventName];
    const idx = handlers.indexOf(callback);
    handlers.splice(idx, 1);
  }

  public findNodeById(id: number) {
    return findNodeById(id);
  }

  public removeAllNode() {
    removeAllNode();
  }
}
