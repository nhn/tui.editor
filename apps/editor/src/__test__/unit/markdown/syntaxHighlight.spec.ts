import { ToastMark } from '@toast-ui/toastmark';
import MarkdownEditor from '@/markdown/mdEditor';
import EventEmitter from '@/event/eventEmitter';
import { source } from 'common-tags';

function getEditorHTML(editor: MarkdownEditor) {
  return editor.view.dom.innerHTML;
}

let mde: MarkdownEditor, em: EventEmitter;

beforeEach(() => {
  em = new EventEmitter();
  mde = new MarkdownEditor(em, { toastMark: new ToastMark() });
});

afterEach(() => {
  mde.destroy();
});

describe('markdown editor syntax highlight', () => {
  it('atx heading', () => {
    mde.setMarkdown('# heading');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('seText heading', () => {
    mde.setMarkdown('heading\n---');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  describe('blockQuote', () => {
    it('basic', () => {
      mde.setMarkdown('> block quote');

      const html = getEditorHTML(mde);

      expect(html).toMatchSnapshot();
    });

    it('with list', () => {
      mde.setMarkdown('> * [ ] block quote');

      const html = getEditorHTML(mde);

      expect(html).toMatchSnapshot();
    });
  });

  it('bulletList', () => {
    mde.setMarkdown('* bullet list');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('orderedList', () => {
    mde.setMarkdown('1. ordered list');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('tastkList', () => {
    mde.setMarkdown('* [x] task list');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  describe('table', () => {
    it('basic', () => {
      const input = source`
          | col2 | col2
          | --- | ---
          | data1 | data2 |
        `;

      mde.setMarkdown(input);

      const html = getEditorHTML(mde);

      expect(html).toMatchSnapshot();
    });

    it('with mark', () => {
      const input = source`
          | col2 | col2
          | --- | ---
          | data1 | **data2** |
        `;

      mde.setMarkdown(input);

      const html = getEditorHTML(mde);

      expect(html).toMatchSnapshot();
    });
  });

  it('thematicBreak', () => {
    mde.setMarkdown('---');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('emph', () => {
    mde.setMarkdown('*emph*');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('strong', () => {
    mde.setMarkdown('**strong**');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('strike', () => {
    mde.setMarkdown('~~strike~~');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('inline code', () => {
    mde.setMarkdown('`inline code`');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('link', () => {
    mde.setMarkdown('[TOAST UI](https://ui.toast.com)');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('image', () => {
    mde.setMarkdown('![Logo](https://picsum.photos/200)');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('code block', () => {
    mde.setMarkdown('```js\nconsole.log("editor")\n```');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('code block within list', () => {
    mde.setMarkdown('* list\n  ```js\n  console.log("editor")\n  ```');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });

  it('custom block', () => {
    mde.setMarkdown('$$custom\nmy custom element\n$$');

    const html = getEditorHTML(mde);

    expect(html).toMatchSnapshot();
  });
});
