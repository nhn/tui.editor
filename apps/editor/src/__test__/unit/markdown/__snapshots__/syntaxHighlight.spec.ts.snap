// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown editor syntax highlight atx heading 1`] = `
<div>
  <span class="toastui-editor-md-heading toastui-editor-md-heading1">
    <span class="toastui-editor-md-delimiter">
      #
    </span>
    heading
  </span>
</div>
`;

exports[`markdown editor syntax highlight blockQuote basic 1`] = `
<div>
  <span class="toastui-editor-md-block-quote">
    &gt;
    <span class="toastui-editor-md-marked-text">
      block quote
    </span>
  </span>
</div>
`;

exports[`markdown editor syntax highlight blockQuote with list 1`] = `
<div>
  <span class="toastui-editor-md-block-quote">
    &gt;
    <span class="toastui-editor-md-list-item toastui-editor-md-list-item-style toastui-editor-md-list-item-odd">
      *
    </span>
    <span class="toastui-editor-md-delimiter toastui-editor-md-list-item">
      [
      <span class="toastui-editor-md-meta">
      </span>
      ]
    </span>
    <span class="toastui-editor-md-marked-text">
      block quote
    </span>
  </span>
</div>
`;

exports[`markdown editor syntax highlight bulletList 1`] = `
<div>
  <span class="toastui-editor-md-list-item toastui-editor-md-list-item-style toastui-editor-md-list-item-odd">
    *
    <span class="toastui-editor-md-marked-text">
    </span>
  </span>
  <span class="toastui-editor-md-marked-text">
    bullet list
  </span>
</div>
`;

exports[`markdown editor syntax highlight code block 1`] = `
<div class="toastui-editor-md-code-block-line-background start">
  <span class="toastui-editor-md-code-block">
    <span class="toastui-editor-md-delimiter">
      \`\`\`
    </span>
    <span class="toastui-editor-md-meta">
      js
    </span>
  </span>
</div>
<div class="toastui-editor-md-code-block-line-background">
  <span class="toastui-editor-md-code-block">
    console.log("editor")
  </span>
</div>
<div class="toastui-editor-md-code-block-line-background">
  <span class="toastui-editor-md-code-block">
    <span class="toastui-editor-md-delimiter">
      \`\`\`
    </span>
  </span>
</div>
`;

exports[`markdown editor syntax highlight code block within list 1`] = `
<div>
  <span class="toastui-editor-md-list-item toastui-editor-md-list-item-style toastui-editor-md-list-item-odd">
    *
    <span class="toastui-editor-md-marked-text">
    </span>
  </span>
  <span class="toastui-editor-md-marked-text">
    list
  </span>
</div>
<div>
  <span class="toastui-editor-md-code-block">
    <span class="toastui-editor-md-marked-text">
    </span>
    <span class="toastui-editor-md-delimiter">
      <span class="toastui-editor-md-marked-text">
        \`\`\`
      </span>
    </span>
    <span class="toastui-editor-md-marked-text">
      <span class="toastui-editor-md-meta">
        js
      </span>
    </span>
  </span>
</div>
<div>
  <span class="toastui-editor-md-code-block">
    <span class="toastui-editor-md-marked-text">
      console.log("editor")
    </span>
  </span>
</div>
<div>
  <span class="toastui-editor-md-code-block">
    <span class="toastui-editor-md-delimiter">
      <span class="toastui-editor-md-marked-text">
        \`\`\`
      </span>
    </span>
  </span>
</div>
`;

exports[`markdown editor syntax highlight custom block 1`] = `
<div class="toastui-editor-md-custom-block-line-background start">
  <span class="toastui-editor-md-custom-block">
    <span class="toastui-editor-md-delimiter">
      $$
    </span>
    <span class="toastui-editor-md-meta">
      custom
    </span>
  </span>
</div>
<div class="toastui-editor-md-custom-block-line-background">
  <span class="toastui-editor-md-custom-block">
    my custom element
  </span>
</div>
<div class="toastui-editor-md-custom-block-line-background">
  <span class="toastui-editor-md-custom-block">
    <span class="toastui-editor-md-delimiter">
      $$
    </span>
  </span>
</div>
`;

exports[`markdown editor syntax highlight emph 1`] = `
<div>
  <span class="toastui-editor-md-delimiter">
    *
  </span>
  <span class="toastui-editor-md-emph">
    emph
  </span>
  <span class="toastui-editor-md-delimiter">
    *
  </span>
</div>
`;

exports[`markdown editor syntax highlight image 1`] = `
<div>
  <span class="toastui-editor-md-link">
    <span class="toastui-editor-md-meta">
      !
    </span>
    [
  </span>
  <span class="toastui-editor-md-link toastui-editor-md-link-desc toastui-editor-md-marked-text">
    Logo
  </span>
  <span class="toastui-editor-md-link">
    ](
  </span>
  <span class="toastui-editor-md-link toastui-editor-md-link-url toastui-editor-md-marked-text">
    https://picsum.photos/200
  </span>
  <span class="toastui-editor-md-link">
    )
  </span>
</div>
`;

exports[`markdown editor syntax highlight inline code 1`] = `
<div>
  <span class="toastui-editor-md-code toastui-editor-md-delimiter toastui-editor-md-start">
    \`
  </span>
  <span class="toastui-editor-md-code toastui-editor-md-marked-text">
    inline code
  </span>
  <span class="toastui-editor-md-code toastui-editor-md-delimiter toastui-editor-md-end">
    \`
  </span>
</div>
`;

exports[`markdown editor syntax highlight link 1`] = `
<div>
  <span class="toastui-editor-md-link">
    [
  </span>
  <span class="toastui-editor-md-link toastui-editor-md-link-desc toastui-editor-md-marked-text">
    TOAST UI
  </span>
  <span class="toastui-editor-md-link">
    ](
  </span>
  <span class="toastui-editor-md-link toastui-editor-md-link-url toastui-editor-md-marked-text">
    https://ui.toast.com
  </span>
  <span class="toastui-editor-md-link">
    )
  </span>
</div>
`;

exports[`markdown editor syntax highlight orderedList 1`] = `
<div>
  <span class="toastui-editor-md-list-item toastui-editor-md-list-item-style toastui-editor-md-list-item-odd">
    1.
    <span class="toastui-editor-md-marked-text">
    </span>
  </span>
  <span class="toastui-editor-md-marked-text">
    ordered list
  </span>
</div>
`;

exports[`markdown editor syntax highlight seText heading 1`] = `
<div>
  <span class="toastui-editor-md-heading toastui-editor-md-heading2">
    heading
  </span>
</div>
<div>
  <span class="toastui-editor-md-heading toastui-editor-md-heading1 toastui-editor-md-delimiter toastui-editor-md-setext">
    ---
  </span>
</div>
`;

exports[`markdown editor syntax highlight strike 1`] = `
<div>
  <span class="toastui-editor-md-delimiter">
    ~~
  </span>
  <span class="toastui-editor-md-strike">
    strike
  </span>
  <span class="toastui-editor-md-delimiter">
    ~~
  </span>
</div>
`;

exports[`markdown editor syntax highlight strong 1`] = `
<div>
  <span class="toastui-editor-md-delimiter">
    **
  </span>
  <span class="toastui-editor-md-strong">
    strong
  </span>
  <span class="toastui-editor-md-delimiter">
    **
  </span>
</div>
`;

exports[`markdown editor syntax highlight table basic 1`] = `
<div>
  <span class="toastui-editor-md-table">
    |
    <span class="toastui-editor-md-table-cell">
      col2
    </span>
    |
    <span class="toastui-editor-md-table-cell">
      col2
    </span>
  </span>
</div>
<div>
  <span class="toastui-editor-md-table">
    | --- | ---
  </span>
</div>
<div>
  <span class="toastui-editor-md-table">
    |
    <span class="toastui-editor-md-table-cell">
      data1
    </span>
    |
    <span class="toastui-editor-md-table-cell">
      data2
    </span>
    |
  </span>
</div>
`;

exports[`markdown editor syntax highlight table with mark 1`] = `
<div>
  <span class="toastui-editor-md-table">
    |
    <span class="toastui-editor-md-table-cell">
      col2
    </span>
    |
    <span class="toastui-editor-md-table-cell">
      col2
    </span>
  </span>
</div>
<div>
  <span class="toastui-editor-md-table">
    | --- | ---
  </span>
</div>
<div>
  <span class="toastui-editor-md-table">
    |
    <span class="toastui-editor-md-table-cell">
      data1
    </span>
    |
    <span class="toastui-editor-md-table-cell">
      <span class="toastui-editor-md-delimiter">
        **
      </span>
      <span class="toastui-editor-md-strong">
        data2
      </span>
      <span class="toastui-editor-md-delimiter">
        **
      </span>
    </span>
    |
  </span>
</div>
`;

exports[`markdown editor syntax highlight tastkList 1`] = `
<div>
  <span class="toastui-editor-md-list-item toastui-editor-md-list-item-style toastui-editor-md-list-item-odd">
    *
  </span>
  <span class="toastui-editor-md-delimiter toastui-editor-md-list-item">
    [
    <span class="toastui-editor-md-meta">
      x
    </span>
    ]
  </span>
  <span class="toastui-editor-md-marked-text">
    task list
  </span>
</div>
`;

exports[`markdown editor syntax highlight thematicBreak 1`] = `
<div>
  <span class="toastui-editor-md-thematic-break">
    ---
  </span>
</div>
`;
